### DISCUSSION TITLE
How to create backup properly when 2FA Yubikeys are being used

### DISCUSSION THREAD

**Discussion by synaptiko** [dani-garcia/vaultwarden#1521] [Q&A]
I'm trying to create a setup with `bitwarden_rs` on my VM and I'd like to have a good way to create automated backups and recovery. Unfortunately after a few tries I found out that if the account has any 2FA with Yubikey associated it doesn't work when I'm using the DB from the backup. I have to reset those. Only normal OTP (with an app) works. Is there something I'm missing in the backup? I follow the wiki and I don't think I'm missing any files.

Also I'm not using `config.json` file, all the config is done through env vars. I'm running BW in the container (with `podman`).

Thanks for any ideas.

**Comment by dumblob** [comment]
Could you share a script and/or YAML file(s) etc. with your whole setup (maybe a link to your repo)? It might be useful as an example setup for others and make reproduction of the issue easier.

**Reply by synaptiko** [reply]
I don't have the script accessible publicly because it contains information related to my system & infra. But I basically followed the instructions on the wiki. To start Bitwarden, I'm using the following command:
```
podman run -d \
	--name bitwarden \
	-p 127.0.0.1:8080:80 \
	-p 127.0.0.1:8081:3012 \
	-v /var/lib/bitwarden/:/data/:Z \
	-e TZ="$TZ" \
	-e LOG_LEVEL=warn \
	-e LOG_FILE="/data/bitwarden.log" \
	-e SHOW_PASSWORD_HINT=false \
	-e ROCKET_LIMITTS={json=104857600} \
	-e DOMAIN="$DOMAIN" \
	-e ADMIN_TOKEN="$ADMIN_TOKEN" \
	-e WEBSOCKET_ENABLED=true \
	-e SIGNUPS_ALLOWED=false \
	-e INVITATIONS_ALLOWED=false \
	-e SMTP_HOST="${SMTP_HOST}" \
	-e SMTP_SSL=true \
	-e SMTP_EXPLICIT_TLS=true \
	-e SMTP_PORT=${SMTP_PORT} \
	-e SMTP_FROM="${SMTP_FROM}" \
	-e SMTP_FROM_NAME="${SMTP_FROM_NAME}" \
	-e SMTP_USERNAME="${SMTP_USERNAME}" \
	-e SMTP_PASSWORD="${SMTP_PASSWORD}" \
	docker.io/bitwardenrs/server:alpine

podman exec -it bitwarden /bin/sh -c "apk add tzdata && cp /usr/share/zoneinfo/$TZ /etc/localtime && echo \"$TZ\" > /etc/timezone"
```

And to backup I'm running:
```
sqlite3 data/db.sqlite3 ".backup '/path/to/backups/db-$(date '+%Y%m%d-%H%M').sqlite3'"
```
inside the container.

**Comment by jjlin** [comment]
All 2FA-related data is stored in the SQL database, so if you've dumped and restored it properly, that's the best you can do. Are you using U2F? That does seem to maintain some state in the database, particularly a device counter:

https://developers.yubico.com/U2F/Protocol_details/Overview.html#_4_device_cloning_detection

But it's not obvious to me why U2F would break after restoring a previous database, since the device should still be providing a counter value higher than what was stored in an older version of the database.

**Reply by synaptiko** [reply]
>  if you've dumped and restored it properly

I believe I did, I followed the instructions on the wiki page.

> Are you using U2F?

Yes, I am.

Is the count just an integer? Do you know what table & column it saved in? I can verify that it's consistent after restore.

**Reply by jjlin** [reply]
You can see it via `sqlite3 data/db.sqlite3 'select * from twofactor'`. Although I just tried restoring from a backup and U2F still works fine for me.

**Reply by synaptiko** [reply]
Thank you, ok, I'll check it once more. Maybe I just messed up something during the backup/recovery.
